### PR TITLE
Typing Core Package

### LINKED ISSUE #651: Improve code with typing

### PR DESCRIPTION
**Pull request by justusschock**
Fixes #651 

Description:
This PR will add typings for the whole core package (`ignite.engine`, `ignite.metrics`, `ignite.handlers) but **NOT** for the contrib part so far.

Check list:
* [x] New tests are added (if a new feature is added) [Not Necessary]
* [x] New doc strings: description and/or example code are in RST format [Not Necessary]
* [x] Documentation is updated (if required) [Not Necessary]

cc @vfdev-5

### CODE REVIEW

**Comment by justusschock** [comment]
The failing tests are just part of the contrib package (which I didn't even touch). How shall we proceed here?

File: ignite/engine/engine.py [hunk 1]
```diff
@@ -182,14 +190,14 @@ def __init__(self, **kwargs):
             if not hasattr(self, value):
                 setattr(self, value, 0)
 
-    def get_event_attrib_value(self, event_name):
+    def get_event_attrib_value(self, event_name: str) -> int:
```
Comment (vfdev-5) [inline ignite/engine/engine.py]: @justusschock I think here and after `event_name` can be `EventWithFilter`, `CallableEvents` or `Enum`

File: ignite/engine/engine.py [hunk 2]
```diff
@@ -227,12 +235,12 @@ def print_epoch(engine):
         # print_epoch handler is now unregistered
     """
 
-    def __init__(self, event_name, handler, engine):
+    def __init__(self, event_name: str, handler: Callable, engine: Engine):
```
Comment (vfdev-5) [inline ignite/engine/engine.py]: Probably, I think `event_name` is not `str` here, but `EventWithFilter`, `CallableEvents` or `Enum`.

File: ignite/handlers/checkpoint.py [hunk 3]
```diff
@@ -392,12 +395,12 @@ def __init__(self, dirname, filename_prefix,
         self.global_step_transform = global_step_transform
 
     @property
-    def last_checkpoint(self):
+    def last_checkpoint(self) -> None:
```
Comment (vfdev-5) [inline ignite/handlers/checkpoint.py]:
```suggestion
    def last_checkpoint(self) -> Union[str, None]:
```

File: ignite/metrics/loss.py [hunk 4]
```diff
@@ -1,4 +1,7 @@
 from __future__ import division
+from typing import Callable, Union, Optional, Sequence
```
Comment (vfdev-5) [inline ignite/metrics/loss.py]: Maybe we can remove above line too as no more python 2:
```python
from __future__ import division
```

File: ignite/metrics/mean_absolute_error.py [hunk 5]
```diff
@@ -1,5 +1,7 @@
 from __future__ import division
```
Comment (vfdev-5) [inline ignite/metrics/mean_absolute_error.py]: Same here.
Maybe we can remove above line too as no more python 2:
```python
from __future__ import division
```

File: ignite/metrics/mean_pairwise_distance.py [hunk 6]
```diff
@@ -1,5 +1,7 @@
 from __future__ import division
```
Comment (vfdev-5) [inline ignite/metrics/mean_pairwise_distance.py]: Same here.

File: ignite/metrics/mean_squared_error.py [hunk 7]
```diff
@@ -1,5 +1,7 @@
 from __future__ import division
```
Comment (vfdev-5) [inline ignite/metrics/mean_squared_error.py]: Same here

File: ignite/metrics/top_k_categorical_accuracy.py [hunk 8]
```diff
@@ -1,5 +1,7 @@
 from __future__ import division
```
Comment (vfdev-5) [inline ignite/metrics/top_k_categorical_accuracy.py]: We probably need to remove this line.

File: ignite/metrics/mean_absolute_error.py (hunk 5 above)
Comment (justusschock) [reply ignite/metrics/mean_absolute_error.py]: Will do this in a separate PR

File: ignite/metrics/mean_squared_error.py (hunk 7 above)
Comment (justusschock) [reply ignite/metrics/mean_squared_error.py]: Will do this in a separate PR

File: ignite/metrics/mean_pairwise_distance.py (hunk 6 above)
Comment (justusschock) [reply ignite/metrics/mean_pairwise_distance.py]: Will do this in a separate PR

File: ignite/metrics/top_k_categorical_accuracy.py (hunk 8 above)
Comment (justusschock) [reply ignite/metrics/top_k_categorical_accuracy.py]: Will do this in a separate PR

File: ignite/metrics/loss.py (hunk 4 above)
Comment (justusschock) [reply ignite/metrics/loss.py]: Will do this in a separate PR

File: ignite/engine/engine.py (hunk 2 above)
Comment (justusschock) [reply ignite/engine/engine.py]: Done

File: ignite/engine/engine.py (hunk 1 above)
Comment (justusschock) [reply ignite/engine/engine.py]: Done

**Comment by vfdev-5** [comment]
@justusschock I let you resolve the conflicts and it will be OK to merge

File: ignite/engine/engine.py [hunk 9]
```diff
@@ -321,6 +323,23 @@ def remove_event_handler(self, handler, event_name):
             raise ValueError("Input handler '{}' is not found among registered event handlers".format(handler))
         self._event_handlers[event_name] = new_event_handlers
 
+    @staticmethod
```
Comment (vfdev-5) [inline ignite/engine/engine.py]: We have now both `_check_signature` in Engine and in utils.py.

Comment (justusschock) [reply ignite/engine/engine.py]: Oh, sorry. Must have missed that during the rebasing. Will fix this!

**Pull request merged**